grakn-graph/src/main/java/ai/grakn/graph/internal/ElementFactory.java

```diff
@@ -42,74 +46,106 @@
+        return buildEntityType(v, Optional.of(type));
+    }
+    private EntityTypeImpl buildEntityType(Vertex v, Optional<EntityType> type){
+        return new EntityTypeImpl(graknGraph, v, type);
     }
 
-    EntityImpl buildEntity(Vertex v, EntityType type){
-        return  new EntityImpl(v, type, graknGraph);
+    // ------------------------------------------- Building Entities
+    public EntityImpl buildEntity(Vertex v, EntityType type){
+        return buildEntity(v, Optional.of(type));
+    }
+    private EntityImpl buildEntity(Vertex v, Optional<EntityType> type){
+        return  new EntityImpl(graknGraph, v, type);
     }
 
-    private <V> ResourceImpl <V> buildResource(Vertex v, ResourceType<V> type){
-        return new ResourceImpl<>(v, type, graknGraph);
+    // ----------------------------------------- Building Rule Types  --------------------------------------------------
+    public RuleTypeImpl buildRuleType(Vertex v, RuleType type){
+        return buildRuleType(v, Optional.of(type));
+    }
+    private RuleTypeImpl buildRuleType(Vertex v, Optional<RuleType> type){
+        return new RuleTypeImpl(graknGraph, v, type);
     }
 
-    <V> ResourceImpl <V> buildResource(Vertex v, ResourceType<V> type, V value){
-        return new ResourceImpl<>(v, type, graknGraph, value);
+    // -------------------------------------------- Building Rules
+    public RuleImpl buildRule(Vertex v, RuleType type, Pattern lhs, Pattern rhs){
+        if(lhs == null)
+            throw new InvalidConceptValueException(ErrorMessage.NULL_VALUE.getMessage(Schema.ConceptProperty.RULE_LHS.name()));
+
+        if(rhs == null)
+            throw new InvalidConceptValueException(ErrorMessage.NULL_VALUE.getMessage(Schema.ConceptProperty.RULE_RHS.name()));
+
+        return buildRule(v, Optional.of(type), Optional.of(lhs), Optional.of(rhs));
+    }
+    private RuleImpl buildRule(Vertex v, Optional<RuleType> type, Optional<Pattern> lhs, Optional<Pattern> rhs){
+        return new RuleImpl(graknGraph, v, type, lhs, rhs);
     }
 
-    private RuleImpl buildRule(Vertex v, RuleType type){
-        Pattern lhs = graknGraph.graql().parsePattern(v.value(Schema.ConceptProperty.RULE_LHS.name()));
-        Pattern rhs = graknGraph.graql().parsePattern(v.value(Schema.ConceptProperty.RULE_RHS.name()));
-        return buildRule(v, type, lhs, rhs);
+    // ------------------------------------------ Building Roles  Types ------------------------------------------------
+    public RoleTypeImpl buildRoleType(Vertex v, RoleType type, Boolean isImplicit){
+        return buildRoleType(v, Optional.of(type), Optional.of(isImplicit));
     }
-    RuleImpl buildRule(Vertex v, RuleType type, Pattern lhs, Pattern rhs){
-        return  new RuleImpl(v, type, graknGraph, lhs, rhs);
+    private RoleTypeImpl buildRoleType(Vertex v, Optional<RoleType> type, Optional<Boolean> isImplicit){
+        return new RoleTypeImpl(graknGraph, v, type, isImplicit);
```

Comment:
I just had a look and I think I was wrong about that! I see a lot of them are used in `buildConcept`, but I don't think those usages can be removed after all.